### ISSUE TITLE
Implement admission webhook for AlertmanagerConfig resources

### ISSUE THREAD

**Issue by simonpasquier** [prometheus-operator/prometheus-operator#3710]
**What is missing?**

Similar to what we have for Prometheus rules, it would be useful to implement an admission webhook endpoint for `AlertmanagerConfig` resources.

**Why do we need it?**

There are validation constraints that can't be modeled in the CRD. To list a few:
* ~~a route needs to reference the name of an existing receiver in the same resource ([link](https://github.com/prometheus-operator/prometheus-operator/blob/fe65bc9ea96270718dd273ea1565d61a7ad3941d/pkg/alertmanager/operator.go#L1091-L1093)).~~ incorrect see #3749 
* Slack actions need to have either an url or a name ([link](https://github.com/prometheus-operator/prometheus-operator/blob/fe65bc9ea96270718dd273ea1565d61a7ad3941d/pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go#L235-L237)).
* sub-routes are only represented as raw JSON data and can only be validated at reconciliation time (#3703).
* url fields need a valid scheme (#4153))

**Comment by sev3ryn** [comment]
@simonpasquier what is your idea around validating of receiver existence used in routes? 
From one perspective we shouldn't validate that - so that some shared receivers specified in main config may be used. From other perspective not validating receivers existence may lead to configuration that alertmanager can't apply

**Comment by sev3ryn** [comment]
I analyzed what needs to be done to add those validations - so basically validation webhook needs to do same job as operator iteself parsing, validating and transforming input to the alertmanager configuration style. 

So maybe to keep code DRY it would be good - to do all the config parsing and transforming job in mutating webhook, write resulting alertmanager configuration to  the annotations. In validating webhook you just check if it unmarshals(and validates) to alertmanager's config.Config. And in operator you will just copy-paste annotation to alertmanagers resulting config file?

**Comment by simonpasquier** [comment]
I'm not sure I'm following your proposal about using annotations. IMHO, the first thing to do would be refactor the code in [pkg/alertmanager/amcfg.go](https://github.com/prometheus-operator/prometheus-operator/blob/master/pkg/alertmanager/amcfg.go) and [pkg/alertmanager/operator.go](https://github.com/prometheus-operator/prometheus-operator/blob/master/pkg/alertmanager/operator.go). For instance, we check the email configurations in both places:

https://github.com/prometheus-operator/prometheus-operator/blob/02a5bac9610299372e9f77cbbe0c824ce636795b/pkg/alertmanager/amcfg.go#L673-L755

https://github.com/prometheus-operator/prometheus-operator/blob/02a5bac9610299372e9f77cbbe0c824ce636795b/pkg/alertmanager/operator.go#L1098-L1140

**Comment by just1900** [comment]
what's the current state of alertmanagerconfig's webhook?
Just running into an issue, here are logs from alertmanager during restart
```
level=error ts=2021-07-07T08:01:25.728Z caller=coordinator.go:124 component=configuration msg="Loading configuration file failed" file=/etc/alertmanager/config/alertmanager.yaml err="unsupported scheme \"\" for URL"
```
As a result, I think we also need to validate the url pattern, otherwise, alertmanager may be CrashLoopBackOff after restart if we misconfigure some alertmanagerconfig without validating.

**Comment by just1900** [comment]
I guess it's better to start a new issue https://github.com/prometheus-operator/prometheus-operator/issues/4153 to trace this problem, rather than discussing here :)

**Comment by just1900** [comment]
As @simonpasquier memtioned, we check configuration in two places for now.
IMHO, we can move the check in `operator.go` to a validatingWebhook(I don't think it a good idea to do another mutatingWebhook and generate config in annotation to be used later) and reserve the basic nil checking logic in `amcfg.go` during config Converting.

**Comment by PhilipGough** [comment]
@simonpasquier Closed by #4388 ?
